Userland/Utilities/readelf.cpp

```diff
@@ -450,23 +578,49 @@ int main(int argc, char** argv)
 
     if (display_dynamic_section) {
         auto found_dynamic_section = false;
-        interpreter_image.for_each_section([&found_dynamic_section](const ELF::Image::Section& section) {
-            if (section.name() != ELF_DYNAMIC)
-                return IterationDecision::Continue;
+        if (elf_image.is_dynamic()) {
+            elf_image.for_each_section([&found_dynamic_section](const ELF::Image::Section& section) {
+                if (section.name() != ELF_DYNAMIC)
+                    return IterationDecision::Continue;
 
-            found_dynamic_section = true;
+                found_dynamic_section = true;
 
-            if (!section.entry_count()) {
-                printf("Dynamic section '%s' at offset 0x%08x contains zero entries.\n", section.name().to_string().characters(), section.offset());
-                return IterationDecision::Continue;
-            }
+                if (section.entry_count()) {
+                    printf("Dynamic section '%s' at offset 0x%08x contains %u entries.\n", section.name().to_string().characters(), section.offset(), section.entry_count());
+                } else {
+                    printf("Dynamic section '%s' at offset 0x%08x contains zero entries.\n", section.name().to_string().characters(), section.offset());
+                }
 
-            printf("Dynamic section '%s' at offset 0x%08x contains %u entries.\n", section.name().to_string().characters(), section.offset(), section.entry_count());
+                return IterationDecision::Break;
+            });
 
-            // TODO: Display dynamic section
+            Vector<String> libraries;
+            object->for_each_needed_library([&libraries](StringView entry) {
+                libraries.append(String::formatted("{}", entry).characters());
+                return IterationDecision::Continue;
+            });
 
-            return IterationDecision::Continue;
-        });
+            auto library_index = 0;
+            printf("  Tag        Type              Name / Value\n");
+            object->for_each_dynamic_entry([&library_index, &libraries](const ELF::DynamicObject::DynamicEntry& entry) {
+                printf("  0x%08x ", entry.tag());
+                printf("%-17s ", object_tag_to_string(entry.tag()));
+
+                if (entry.tag() == DT_NEEDED) {
+                    printf("Shared library: %s\n", String(libraries[library_index]).characters());
+                    library_index++;
+                } else if (entry.tag() == DT_RPATH) {
+                    // FIXME: print the rpath
+                    printf("Library rpath: 0x%08x\n", entry.val());
```

Comment:
Should also be able to print SONAME here-- there's a helper on Dynamic object to get it from the .dynstr section. The same strategy for soname() could be used for runpath/rpath.

Reply:
I've added LibELF helpers for `rpath` and `runpath` and added these and `soname` to readelf for now.

The helpers will lie in the event of duplicate entries.

Using helpers isn't ideal. Eventually `readelf` should resolve entries directly via offset rather than relying on helpers. There's still a lot work to be done on LibELF.